### PR TITLE
ARROW-16817: [C++][Python] Fix segfaults in the ORC writer when unsupported datatypes are given

### PR DESCRIPTION
**Pull request by iajoiner**

### CODE REVIEW

**Comment by iajoiner** [comment]
Ok looks like the bug no longer exists in Arrow 8.0.0 likely due to the fix to https://issues.apache.org/jira/browse/ARROW-16025. These additional tests in C++ & Python are still good to add though just in case future updates break them.

**Comment by iajoiner** [comment]
All tests have passed except for two that have nothing to do with the PR. Please review.

File: cpp/src/arrow/adapters/orc/adapter_test.cc [hunk 1]
```diff
@@ -487,6 +487,36 @@ TEST_F(TestORCWriterTrivialWithConversion, writeChunkless) {
...
+ public:
+  TestORCWriterInvalidTypes() {
+    table_schema = schema({field("uint8", uint8()), field("uint16", uint16()),
+                           field("uint32", uint32()), field("uint64", uint64())});
+    auto write_options = adapters::orc::WriteOptions();
+#ifdef ARROW_WITH_SNAPPY
+    write_options.compression = Compression::SNAPPY;
+#else
+    write_options.compression = Compression::UNCOMPRESSED;
+#endif
```
Comment (pitrou) [inline cpp/src/arrow/adapters/orc/adapter_test.cc:501]: This doesn't seem necessary, is it?

File: cpp/src/arrow/adapters/orc/adapter_test.cc [hunk 2]
```diff
@@ -487,6 +487,36 @@ TEST_F(TestORCWriterTrivialWithConversion, writeChunkless) {
...
+    write_options.compression_block_size = 32768;
+    write_options.row_index_stride = 5000;
+  }
+
+ protected:
+  std::shared_ptr<Schema> table_schema;
+  adapters::orc::WriteOptions write_options;
+};
+TEST_F(TestORCWriterInvalidTypes, noWriteInvalidTypes) {
+  const std::shared_ptr<Table> table = GenerateRandomTable(table_schema, 5332, 5, 7, 0);
```
Comment (pitrou) [inline cpp/src/arrow/adapters/orc/adapter_test.cc:512]: Why 5332 rows? I would expect the error to show up even with a small number of rows. There's no reason to generate a lot of data here.

File: cpp/src/arrow/adapters/orc/adapter_test.cc [hunk 3]
```diff
@@ -487,6 +487,36 @@ TEST_F(TestORCWriterTrivialWithConversion, writeChunkless) {
...
+    write_options.file_version = adapters::orc::FileVersion(0, 11);
+    write_options.compression_block_size = 32768;
+    write_options.row_index_stride = 5000;
+  }
+
+ protected:
+  std::shared_ptr<Schema> table_schema;
+  adapters::orc::WriteOptions write_options;
+};
+TEST_F(TestORCWriterInvalidTypes, noWriteInvalidTypes) {
```
Comment (pitrou) [inline cpp/src/arrow/adapters/orc/adapter_test.cc:511]: The test name is cryptic. How are the types invalid? Why "noWrite"?

File: cpp/src/arrow/adapters/orc/adapter_test.cc [hunk 4]
```diff
@@ -487,6 +487,36 @@ TEST_F(TestORCWriterTrivialWithConversion, writeChunkless) {
                             kDefaultSmallMemStreamSize / 16);
 }
 
+// Invalid Types
+class TestORCWriterInvalidTypes : public ::testing::Test {
+ public:
+  TestORCWriterInvalidTypes() {
```
Comment (pitrou) [inline cpp/src/arrow/adapters/orc/adapter_test.cc:493]: Usually this would be a `SetUp()` method instead.

Comment (pitrou) [reply cpp/src/arrow/adapters/orc/adapter_test.cc:493]:
```suggestion
  void SetUp() override {
```

File: cpp/src/arrow/adapters/orc/adapter_test.cc [hunk 5]
```diff
@@ -487,6 +487,36 @@ TEST_F(TestORCWriterTrivialWithConversion, writeChunkless) {
...
+    table_schema = schema({field("uint8", uint8()), field("uint16", uint16()),
+                           field("uint32", uint32()), field("uint64", uint64())});
+    auto write_options = adapters::orc::WriteOptions();
```
Comment (pitrou) [inline cpp/src/arrow/adapters/orc/adapter_test.cc:496]: `auto` declares a local variable, so this won't set the member variable.

File: cpp/src/arrow/adapters/orc/adapter_test.cc [hunk 6]
```diff
@@ -487,6 +487,36 @@ TEST_F(TestORCWriterTrivialWithConversion, writeChunkless) {
...
+    write_options.compression = Compression::UNCOMPRESSED;
+#endif
+    write_options.file_version = adapters::orc::FileVersion(0, 11);
+    write_options.compression_block_size = 32768;
+    write_options.row_index_stride = 5000;
+  }
+
+ protected:
+  std::shared_ptr<Schema> table_schema;
+  adapters::orc::WriteOptions write_options;
```
Comment (pitrou) [inline cpp/src/arrow/adapters/orc/adapter_test.cc:509]:
```suggestion
  std::shared_ptr<Schema> table_schema_;
  adapters::orc::WriteOptions write_options_;
```

File: cpp/src/arrow/adapters/orc/adapter_test.cc [hunk 7]
```diff
@@ -487,6 +487,36 @@ TEST_F(TestORCWriterTrivialWithConversion, writeChunkless) {
...
+};
+TEST_F(TestORCWriterInvalidTypes, noWriteInvalidTypes) {
+  const std::shared_ptr<Table> table = GenerateRandomTable(table_schema, 5332, 5, 7, 0);
+  EXPECT_OK_AND_ASSIGN(auto buffer_output_stream,
+                       io::BufferOutputStream::Create(kDefaultSmallMemStreamSize / 16));
+  EXPECT_OK_AND_ASSIGN(auto writer, adapters::orc::ORCFileWriter::Open(
+                                        buffer_output_stream.get(), write_options));
+  ASSERT_RAISES(NotImplemented, writer->Write(*table));
```
Comment (pitrou) [inline cpp/src/arrow/adapters/orc/adapter_test.cc:517]: Also, for clarity, this should also check the message error (or some significant part thereof). You can look for instances of `EXPECT_FINISHES_AND_RAISES_WITH_MESSAGE_THAT` for examples.

File: cpp/src/arrow/adapters/orc/util.cc [hunk 8]
```diff
@@ -897,20 +897,6 @@ Result<ORC_UNIQUE_PTR<liborc::Type>> GetOrcType(const DataType& type) {
       ARROW_ASSIGN_OR_RAISE(auto item_orc_type, GetOrcType(*item_arrow_type));
       return liborc::createMapType(std::move(key_orc_type), std::move(item_orc_type));
     }
-    case Type::type::DENSE_UNION:
-    case Type::type::SPARSE_UNION: {
```
Comment (pitrou) [inline cpp/src/arrow/adapters/orc/util.cc:901]: Why are these being removed? Is this addressing the same issue?

File: python/pyarrow/tests/test_orc.py [hunk 9]
```diff
@@ -626,11 +626,20 @@ def test_wrong_usage_orc_writer(tempdir):
...
 
     path = str(tempdir / 'test.orc')
     a = pa.array([1, None, 3, None])
     b = pa.array([None, None, None, None])
     table = pa.table({"int64": a, "utf8": b})
     with pytest.raises(pa.ArrowNotImplementedError):
         orc.write_table(table, path)
+
+
+def test_orc_writer_with_arrays_with_unsupported_types():
```
Comment (pitrou) [inline python/pyarrow/tests/test_orc.py:638]: There's a test on the C++ side, I don't think it's worth checking it a second time here. These tests should probably focus on exercising the Python APIs, not reproduce the existing C++ tests.

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for doing this. This is probably a welcome addition; see some comments below.

**Comment by pitrou** [comment]
@iajoiner Ping.

**Comment by iajoiner** [comment]
@pitrou Sorry yes. I’m mostly on the Rust side now but I will finish this one in a week.

**Comment by pitrou** [comment]
@iajoiner Looking forward to it! (and glad to hear the Arrow Rust community is ever growing :-))

**Comment by pitrou** [comment]
@iajoiner Are you willing to finish this?

**Comment by wjones127** [comment]
@pitrou Are we still trying to get this into 10.0.0? Or can this be bumped to 11.0.0?

I can work on this regardless.

**Comment by pitrou** [comment]
It can probably be bumped IMHO.

**Comment by wjones127** [comment]
Closing in favor of #14638

**Pull request closed without merging**